File: Estruturais/Composite.py
# Componente básico
class Componente:
    def operar(self):
        pass

# Folha (um item indivisível)
class Folha(Componente):
    def operar(self):
        print("Folha operando")

# Galho (pode conter outros componentes)
class Galho(Componente):
    def __init__(self):
        self.componentes = []

    def adicionar(self, componente):
        self.componentes.append(componente)

    def operar(self):
        print("Galho operando")
        for componente in self.componentes:
            componente.operar()

# Exemplo de uso: uma árvore com galhos e folhas
galho1 = Galho()
folha1 = Folha()
galho1.adicionar(folha1)

galho2 = Galho()
folha2 = Folha()
galho2.adicionar(folha2)

galhoPrincipal = Galho()
galhoPrincipal.adicionar(galho1)
galhoPrincipal.adicionar(galho2)

galhoPrincipal.operar()

File: Estruturais/Composite.java
package Estruturais;

import java.util.ArrayList;
import java.util.List;

// Componente
interface Componente {
    void operar();
}

// Folha
class Folha implements Componente {
    public void operar() {
        System.out.println("Folha operando");
    }
}

// Galho
class Galho implements Componente {
    private List<Componente> componentes = new ArrayList<>();

    public void adicionar(Componente componente) {
        componentes.add(componente);
    }

    public void operar() {
        System.out.println("Galho operando");
        for (Componente componente : componentes) {
            componente.operar();
        }
    }
}

public class Composite {
    public static void main(String[] args) {
        Galho galho1 = new Galho();
        Folha folha1 = new Folha();
        galho1.adicionar(folha1);

        Galho galho2 = new Galho();
        Folha folha2 = new Folha();
        galho2.adicionar(folha2);

        Galho galhoPrincipal = new Galho();
        galhoPrincipal.adicionar(galho1);
        galhoPrincipal.adicionar(galho2);

        galhoPrincipal.operar();  // Galho operando, Folha operando (para cada folha)
    }
}

